### PR TITLE
Fix/Rework Bookstack Update function

### LINKED ISSUE #757: BookStack script leaves users in a questionable state in regards to updates
### Please verify that you have read and understood the guidelines.

yes

### A clear and concise description of the issue.

Currently [the script](https://github.com/community-scripts/ProxmoxVE/blob/e4354c1d53ca37eab38d898a33644241d072d8a8/install/bookstack-install.sh#L48-L50) is downloading BookStack from its source ZIP on GitHub, then unzipping that into a `/opt/bookstack` directory.

When it comes to updates, I'm not sure what's expected to occur, but the update script is getting the new version ZIP from GitHub, then extracting this before moving it into the /opt/bookstack directory:

https://github.com/community-scripts/ProxmoxVE/blob/e4354c1d53ca37eab38d898a33644241d072d8a8/ct/bookstack.sh#L69-L71

From my understanding, and later test of the script, this results in the new version files just being moved within the existing install (for example, at `/opt/bookstack/BookStack-24.10.3` if going between the latest releases) so does not actually result in an update.

Based upon how the `.env` file is handled, i assume that the intent is for the original install folder to be wiped out first then replaced, (which would change the behaviour of the `mv` to make sense) but this would be dangerous if intended because I can't see any handling of other potential user files (Uploaded images, customizations, attachments etc...).
Maybe the intent is some kind of merge/overwrite, but then I can't see how core app file/folder removals would be handled as expected.

Generally, the approach taken (manual handling of the source ZIP archives) is not an approach that's officially supported in BookStack, nor is common or expected, so attempts like this need careful consideration. The official guidance currently uses git to manage the core code-base.

For context, I'm the maintainer of BookStack and have encountered this while attempting to support a user to update.

### What settings are you currently utilizing?

- [ ] Default Settings
- [X] Advanced Settings

### Which Linux distribution are you employing?

Debian 12

### If relevant, including screenshots or a code block can be helpful in clarifying the issue.

_No response_

### Please provide detailed steps to reproduce the issue.

_No response_

### PR DESCRIPTION
**Pull request by michelroegl-brunner**
> [!NOTE]
> We are meticulous when it comes to merging code into the main branch, so please understand that we may reject pull requests that do not meet the project's standards. It's never personal. Also, game-related scripts have a lower chance of being merged.

## Description

Complete Rework of the Update function, so it actually updates the APP.
It now creates a backup of all relevant files in an archive, as well as a dump of the Database. 
I
Fixes #757

## Type of change
Please check the relevant option(s):

- [X] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [ ] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [X] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [ ] Testing performed (I have tested my changes, ensuring everything works as expected)
- [ ] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)
Provide any additional context or screenshots about the feature or fix here.

## Related Pull Requests / Discussions

If there are other pull requests or discussions related to this change, please link them here:
- Related PR #

### CODE REVIEW

File: ct/bookstack.sh [hunk 1]
```diff
@@ -60,30 +60,45 @@ check_container_resources
 if [[ ! -d /opt/bookstack ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
 RELEASE=$(curl -s https://api.github.com/repos/BookStackApp/BookStack/releases/latest | grep "tag_name" | awk '{print substr($2, 3, length($2)-4) }')
 if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then
+
   msg_info "Stopping Apache2"
-  systemctl stop apache2
+  systemctl stop apache2t
```
Comment (MickLesk) [inline ct/bookstack.sh:65]: t? ^^

File: ct/bookstack.sh [hunk 2]
```diff
@@ -60,30 +60,45 @@ check_container_resources
...
 if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then
+
   msg_info "Stopping Apache2"
-  systemctl stop apache2
+  systemctl stop apache2t 
   msg_ok "Services Stopped"
 
   msg_info "Updating ${APP} to ${RELEASE}"
-  cp /opt/bookstack/.env /opt/.env
+  cp -r /opt/bookstack/ /opt/bookstack-backup
```
Comment (MickLesk) [inline ct/bookstack.sh:69]: mv better?

File: ct/bookstack.sh [hunk 3]
```diff
@@ -60,30 +60,45 @@ check_container_resources
...
   msg_info "Stopping Apache2"
-  systemctl stop apache2
+  systemctl stop apache2t 
   msg_ok "Services Stopped"
 
   msg_info "Updating ${APP} to ${RELEASE}"
-  cp /opt/bookstack/.env /opt/.env
+  cp -r /opt/bookstack/ /opt/bookstack-backup 
+  tar -czf /opt/bookstack_backup_${RELEASE}.tar.gz /opt/bookstack/.env /opt/bookstack/public/uploads /opt/bookstack/storage/uploads /opt/bookstack/themes &>/dev/null
+  mysqldump -u root bookstack > /opt/bookstack_backup_${RELEASE}.sql
```
Comment (MickLesk) [inline ct/bookstack.sh:71]: here bookstack_backup, above bookstack-backup?

File: ct/bookstack.sh [hunk 4]
```diff
@@ -60,30 +60,45 @@ check_container_resources
...
   php artisan migrate --force &>/dev/null
+  chown www-data:www-data -R /opt/bookstack /opt/bookstack/bootstrap/cache /opt/bookstack/public/uploads /opt/bookstack/storage 
+  chmod -R 755 /opt/bookstack /opt/bookstack/bootstrap/cache /opt/bookstack/public/uploads /opt/bookstack/storage 
+  chmod -R 775 /opt/bookstack/storage /opt/bookstack/bootstrap/cache /opt/bookstack/public/uploads
+  chmod -R 640 /opt/bookstack/.env 
   echo "${RELEASE}" >/opt/${APP}_version.txt
   msg_ok "Updated ${APP}"
 
-  msg_info "Starting Apache2"
+  msg_info "Starting Apache2 "
```
Comment (MickLesk) [inline ct/bookstack.sh:91]: remove space

File: ct/bookstack.sh [hunk 5]
```diff
@@ -60,30 +60,45 @@ check_container_resources
...
   systemctl start apache2
   msg_ok "Started Apache2"
 
   msg_info "Cleaning Up"
-  rm -rf v${RELEASE}.zip
-  msg_ok "Cleaned"
+  cd /root/
```
Comment (MickLesk) [inline ct/bookstack.sh:96]: not needed?

File: ct/bookstack.sh (hunk 1 above)
Comment (michelroegl-brunner) [reply ct/bookstack.sh:65]: yeah, thanks, fixed. sometimes i have neovim^^

File: ct/bookstack.sh (hunk 2 above)
Comment (michelroegl-brunner) [reply ct/bookstack.sh:69]: we can, but we delete /opt/bookstack anyway.

File: ct/bookstack.sh (hunk 3 above)
Comment (michelroegl-brunner) [reply ct/bookstack.sh:71]: changed tar and sql to `bookstack-backup`

File: ct/bookstack.sh (hunk 5 above)
Comment (michelroegl-brunner) [reply ct/bookstack.sh:96]: It is needed, we swicht to `cd /opt/bookstack` above. 
But i can remove it and change 
```
rm -rf /opt/bookstack-backup
rm -rf ${RELEASE}.zip
rm -rf BookStack-${RELEASE}
```
to
```
rm -rf /opt/bookstack-backup
rm -rf /root/${RELEASE}.zip
rm -rf /root/BookStack-${RELEASE}
```
But i find the above more descriptive.

Comment (MickLesk) [reply ct/bookstack.sh:96]: Or rm -rf ~/Filename ?

Comment (michelroegl-brunner) [reply ct/bookstack.sh]: Yeah, i like that. Thank you!

**Review by MickLesk — DISMISSED** [review]

File: ct/bookstack.sh [hunk 6]
```diff
@@ -60,20 +60,31 @@ check_container_resources
...
+
   msg_info "Stopping Apache2"
-  systemctl stop apache2
+  systemctl stop apache2 
   msg_ok "Services Stopped"
 
   msg_info "Updating ${APP} to ${RELEASE}"
-  cp /opt/bookstack/.env /opt/.env
+  mv /opt/bookstack/ /opt/bookstack-backup 
+  tar -czf /opt/bookstack-backup_${RELEASE}.tar.gz /opt/bookstack/.env /opt/bookstack/public/uploads /opt/bookstack/storage/uploads /opt/bookstack/themes &>/dev/null
```
Comment (havardthom) [inline ct/bookstack.sh:70]:
```suggestion
```
I don't think we should start mixing actual backup strategies with zip files and dumps into our update scripts. A temporary backup folder while updating which is used for migrating and is removed afterwards is enough for our scope imo. Actual backups should be handles by users themselves, by having LXC backups/snaphots.

File: ct/bookstack.sh [hunk 7]
```diff
@@ -60,20 +60,31 @@ check_container_resources
...
+  mysqldump -u root bookstack > /opt/bookstack-backup_${RELEASE}.sql
+  rm -rf /opt/bookstack/*
```
Comment (havardthom) [inline ct/bookstack.sh:72]:
```suggestion
  rm -rf /opt/bookstack
```

File: ct/bookstack.sh [hunk 8]
```diff
@@ -60,20 +60,31 @@ check_container_resources
...
   wget -q "https://github.com/BookStackApp/BookStack/archive/refs/tags/v${RELEASE}.zip"
   unzip -q v${RELEASE}.zip
-  mv BookStack-${RELEASE} /opt/bookstack
-  mv /opt/.env /opt/bookstack/.env
+  mv BookStack-${RELEASE}/* /opt/bookstack
```
Comment (havardthom) [inline ct/bookstack.sh:75]:
```suggestion
  mv BookStack-${RELEASE} /opt/bookstack
```

File: ct/bookstack.sh [hunk 9]
```diff
@@ -82,8 +93,11 @@ if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_v
   msg_ok "Started Apache2"
 
   msg_info "Cleaning Up"
-  rm -rf v${RELEASE}.zip
-  msg_ok "Cleaned"
+  rm -rf /opt/bookstack-backup
+  rm -rf ~/${RELEASE}.zip
+  rm -rf ~/BookStack-${RELEASE}
```
Comment (havardthom) [inline ct/bookstack.sh:98]:
```suggestion
```
don't think this exists after being moved

File: ct/bookstack.sh [hunk 10]
```diff
@@ -60,20 +60,31 @@ check_container_resources
 if [[ ! -d /opt/bookstack ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
 RELEASE=$(curl -s https://api.github.com/repos/BookStackApp/BookStack/releases/latest | grep "tag_name" | awk '{print substr($2, 3, length($2)-4) }')
 if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then
+
   msg_info "Stopping Apache2"
-  systemctl stop apache2
+  systemctl stop apache2
```
Comment (havardthom) [inline ct/bookstack.sh:65]:
```suggestion
  systemctl stop apache2
```

File: ct/bookstack.sh [hunk 11]
```diff
@@ -60,20 +60,31 @@ check_container_resources
...
 if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then
+
   msg_info "Stopping Apache2"
-  systemctl stop apache2
+  systemctl stop apache2 
   msg_ok "Services Stopped"
 
   msg_info "Updating ${APP} to ${RELEASE}"
-  cp /opt/bookstack/.env /opt/.env
+  mv /opt/bookstack/ /opt/bookstack-backup
```
Comment (havardthom) [inline ct/bookstack.sh:69]:
```suggestion
  mv /opt/bookstack /opt/bookstack-backup 
```

**Review by havardthom — COMMENTED** [review]
Should be tested thoroughly, with successful data migration of settings/uploads/themes

File: ct/bookstack.sh (hunk 6 above)
Comment (michelroegl-brunner) [reply ct/bookstack.sh:70]: Yeah you a prob right.  I just stuck to the guide there and thought it looks like a good idea to implement that. Maybe I think about this a bit more and start a discussion about this topic.

I will also remove the sql dump then.

**Comment by michelroegl-brunner** [comment]
I´m on a Business trip until Friday evening. I will work in the changes when I get back home. 
I have testet the file restore beforehand, I test it again after the changes.

**Comment by michelroegl-brunner** [comment]
I have made the requested changes and removed the backup part. I testet it and the data migration seems to be working. 
If you can, please test it also if you can: bash -c "$(wget -qLO - https://raw.githubusercontent.com/michelroegl-brunner/ProxmoxVE/refs/heads/dev/ct/bookstack.sh)"
Thank you!

File: ct/bookstack.sh [hunk 12]
```diff
@@ -63,33 +63,40 @@ if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_v
   msg_info "Stopping Apache2"
   systemctl stop apache2
   msg_ok "Services Stopped"
-
   msg_info "Updating ${APP} to ${RELEASE}"
```
Comment (havardthom) [inline ct/bookstack.sh:66]:
```suggestion
  msg_info "Updating ${APP} to v${RELEASE}"
```

File: ct/bookstack.sh [hunk 13]
```diff
@@ -63,33 +63,40 @@ if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_v
...
-  cp /opt/bookstack/.env /opt/.env
+  mv /opt/bookstack /opt/bookstack-backup 
   wget -q "https://github.com/BookStackApp/BookStack/archive/refs/tags/v${RELEASE}.zip"
   unzip -q v${RELEASE}.zip
-  mv BookStack-${RELEASE} /opt/bookstack
-  mv /opt/.env /opt/bookstack/.env
+  mv BookStack-${RELEASE}/ /opt/bookstack
```
Comment (havardthom) [inline ct/bookstack.sh:70]:
```suggestion
  mv BookStack-${RELEASE} /opt/bookstack
```

File: ct/bookstack.sh [hunk 14]
```diff
@@ -63,33 +63,40 @@ if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_v
...
   msg_ok "Updated ${APP}"
-
   msg_info "Starting Apache2"
   systemctl start apache2
   msg_ok "Started Apache2"
-
   msg_info "Cleaning Up"
-  rm -rf v${RELEASE}.zip
-  msg_ok "Cleaned"
+  cd /root/
```
Comment (havardthom) [inline ct/bookstack.sh:89]: Do you know that the current working directory is `/root/` when starting the update script? What if the script is started in some other directory?

File: ct/bookstack.sh [hunk 15]
```diff
@@ -63,33 +63,40 @@ if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_v
...
+  rm -rf /opt/bookstack-backup
+  rm -rf ${RELEASE}.zip
```
Comment (havardthom) [inline ct/bookstack.sh:91]:
```suggestion
  rm -rf v${RELEASE}.zip
```

File: ct/bookstack.sh [hunk 16]
```diff
@@ -63,33 +63,40 @@ if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_v
...
+  msg_ok "Cleaned" 
   msg_ok "Updated Successfully"
 else
   msg_ok "No update required. ${APP} is already at ${RELEASE}"
```
Comment (havardthom) [inline ct/bookstack.sh:95]:
```suggestion
  msg_ok "No update required. ${APP} is already at v${RELEASE}"
```

File: ct/bookstack.sh (hunk 14 above)
Comment (michelroegl-brunner) [reply ct/bookstack.sh:89]: I change it so it is ensured the files we clean are always at /root/. The need is, I switch to /opt/bookstack before the cleanup.

**Pull request closed without merging**